nemo/collections/common/data/dataset.py

```diff
@@ -286,3 +289,310 @@ def __len__(self):
+            if self.shuffle:
+                np.random.shuffle(indices)
+            return iter(indices)
+
+    def build_single_CS_sample(self):
+        comp_text = torch.LongTensor([])
+        created_sample_duration_sec = 0
+        created_sample_langs = []
+        created_sample_audios = []
+
+        pure_mono = np.random.rand() <= self.min_monolingual
+
+        while created_sample_duration_sec < self.min_duration:
+            if (self.pure_random and not pure_mono) or (
+                len(set(created_sample_langs)) == 0 or len(set(created_sample_langs)) == len(self.langs)
+            ):
+                lang_id = np.random.choice(self.langs, p=self.lang_probs)
+            # elif pure_mono:
+            #    use this approach if you want synthetic utterances which are all monolingual
+            #    lang_id = created_sample_langs[0]
+            else:
+                p = np.array(list(map(self.prob_dict.get, list(self.langs_set - set(created_sample_langs)))))
+                p = p / p.sum()
+                lang_id = np.random.choice(list(self.langs_set - set(created_sample_langs)), p=p)
+
+            audio, audio_len, labels, labels_len, *_ = self.get_sample_from_language(lang_id)
+
+            if audio.count_nonzero().item() == 0:
+                continue
+
+            sample_duration = len(audio) / self.sample_rate
+            if (created_sample_duration_sec + sample_duration) > self.max_duration:
+                continue
+
+            if comp_text.device != labels.device:
+                comp_text = comp_text.to(labels.device)
+
+            if audio.ndim > 1 and self.force_monochannel:
+                audio = audio.mean(dim=-1)
+
+            created_sample_duration_sec += sample_duration
+            created_sample_langs.append(lang_id)
+            # need to use numpy instead of torch here because we need numpy's trim_zeros function
+            created_sample_audios.append(audio.cpu().numpy())
+            comp_text = torch.cat([comp_text, labels], dim=0)
+
+            if pure_mono:
+                break
+
+        sample_channels = list(set([s.ndim for s in created_sample_audios]))
+        if len(sample_channels) > 1:
+            raise RuntimeError(
+                "Mixture of audios with different number of channels in CodeSwitchedDataset. All sources must be same number of channels."
+            )
+
+        multichannel = sample_channels[0] > 1
+
+        if multichannel:
+            comp_audio = np.zeros(
+                shape=(int(self.pause_start * self.sample_rate / 1000.0), created_sample_audios[0].shape[-1]),
+                dtype=created_sample_audios[0].dtype,
+            )
+        else:
+            comp_audio = np.zeros(
+                shape=(int(self.pause_start * self.sample_rate / 1000.0),), dtype=created_sample_audios[0].dtype
+            )
+
+        for idx, wav in enumerate(created_sample_audios):
+            if not multichannel:
+                wav = np.trim_zeros(wav)
```

Comment:
The aim here is to remove leading and trailing silences from the sample right? In that case would it be better to achieve this using VAD and a low db threshold for the case where silence/noise is of small amplitude but non zero?

Reply:
It already takes quite a bit of processing time to create the synthetic utterances on-the-fly, so I think VAD would only complicate that further. It's a good idea to revisit for a V2 approach though, perhaps if we can figure out how to do all the processing on GPU if one is detected.

Reply:
VAD is too expensive in a data loader + complicates data parallel with multiple threads (multiple CUDA instances cannot be run on threads, only sub-processes). 
User should do it outside of training.